libs/muddle/internal/kademlia/peer_tracker.hpp

```diff
@@ -0,0 +1,232 @@
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "blacklist.hpp"
+#include "core/byte_array/const_byte_array.hpp"
+#include "core/periodic_runnable.hpp"
+#include "core/reactor.hpp"
+#include "core/service_ids.hpp"
+#include "kademlia/address_priority.hpp"
+#include "kademlia/peer_tracker_protocol.hpp"
+#include "kademlia/table.hpp"
+#include "muddle/rpc/client.hpp"
+#include "muddle/rpc/server.hpp"
+#include "muddle/tracker_configuration.hpp"
+#include "muddle_register.hpp"
+#include "peer_list.hpp"
+#include "promise_runnable.hpp"
+
+#include <chrono>
+#include <memory>
+#include <queue>
+#include <vector>
+
+namespace fetch {
+namespace muddle {
+
+class Muddle;
+class PeerTracker : public core::PeriodicRunnable
+{
+public:
+  using Clock                  = std::chrono::steady_clock;
+  using TimePoint              = Clock::time_point;
+  using Address                = Packet::Address;
+  using Peers                  = std::deque<PeerInfo>;
+  using PeerTrackerPtr         = std::shared_ptr<PeerTracker>;
+  using PeerList               = std::unordered_set<Address>;
+  using PendingPortResolution  = std::unordered_map<Address, std::shared_ptr<PromiseTask>>;
+  using PendingPromised        = std::unordered_map<uint64_t, std::shared_ptr<PromiseTask>>;
+  using ConnectionHandle       = network::AbstractConnection::ConnectionHandleType;
+  using ConstByteArray         = byte_array::ConstByteArray;
+  using PortsList              = PeerTrackerProtocol::PortsList;
+  using ConnectionPriorityMap  = std::unordered_map<Address, AddressPriority>;
+  using ConnectionPriorityList = std::vector<AddressPriority>;
+  using AddressSet             = std::unordered_set<Address>;
+  using AddressMap             = std::unordered_map<Address, Address>;
+  using AddressTimestamp       = std::unordered_map<Address, TimePoint>;
+  using PeerInfoList           = std::deque<PeerInfo>;
+  using BlackList              = fetch::muddle::Blacklist;
+
+  struct UnresolvedConnection
+  {
+    ConnectionHandle handle;
+    Address          address{};
+    std::string      partial_uri{};
+    PortsList        ports{};
+  };
+
+  enum ConnectionState
+  {
+    WAITING  = 1,
+    RESOLVED = 2,
+    DEAD     = 3
+  };
+
+  static PeerTrackerPtr New(Duration const &interval, core::Reactor &reactor,
+                            MuddleRegister const &reg, PeerConnectionList &connections,
+                            MuddleEndpoint &endpoint);
+
+  /// Tracker interface
+  /// @{
+  AddressSet GetDesiredPeers() const;
+  void       AddDesiredPeer(Address const &address);
+  void       AddDesiredPeer(Address const &address, network::Peer const &hint);
+  void       RemoveDesiredPeer(Address const &address);
+  // TODO(tfr): Address GetRoutingAddress(Address const& destination);
```

Comment:
Implement and add `GetRoutingHandle -> {uint64_t, bool}` where uint64_t is handle and bool is whether it is direct or not.